File: bindings/lua/Polycode/SceneMesh.lua
require "Polycode/Entity"

class "SceneMesh" (Entity)


function SceneMesh:__getvar(name)
	if name == "lineWidth" then
		return Polycode.SceneMesh_get_lineWidth(self.__ptr)
	elseif name == "lineSmooth" then
		return Polycode.SceneMesh_get_lineSmooth(self.__ptr)
	elseif name == "pointSmooth" then
		return Polycode.SceneMesh_get_pointSmooth(self.__ptr)
	elseif name == "useGeometryHitDetection" then
		return Polycode.SceneMesh_get_useGeometryHitDetection(self.__ptr)
	elseif name == "alphaTest" then
		return Polycode.SceneMesh_get_alphaTest(self.__ptr)
	elseif name == "backfaceCulled" then
		return Polycode.SceneMesh_get_backfaceCulled(self.__ptr)
	elseif name == "sendBoneMatricesToMaterial" then
		return Polycode.SceneMesh_get_sendBoneMatricesToMaterial(self.__ptr)
	end
	if Entity["__getvar"] ~= nil then
		return Entity.__getvar(self, name)
	end
end

function SceneMesh:__setvar(name,value)
	if name == "lineWidth" then
		Polycode.SceneMesh_set_lineWidth(self.__ptr, value)
		return true
	elseif name == "lineSmooth" then
		Polycode.SceneMesh_set_lineSmooth(self.__ptr, value)
		return true
	elseif name == "pointSmooth" then
		Polycode.SceneMesh_set_pointSmooth(self.__ptr, value)
		return true
	elseif name == "useGeometryHitDetection" then
		Polycode.SceneMesh_set_useGeometryHitDetection(self.__ptr, value)
		return true
	elseif name == "alphaTest" then
		Polycode.SceneMesh_set_alphaTest(self.__ptr, value)
		return true
	elseif name == "backfaceCulled" then
		Polycode.SceneMesh_set_backfaceCulled(self.__ptr, value)
		return true
	elseif name == "sendBoneMatricesToMaterial" then
		Polycode.SceneMesh_set_sendBoneMatricesToMaterial(self.__ptr, value)
		return true
	end
	if Entity["__setvar"] ~= nil then
		return Entity.__setvar(self, name, value)
	else
		return false
	end
end
function SceneMesh:SceneMesh(...)
	local arg = {...}
	if type(arg[1]) == "table" and count(arg) == 1 then
		if ""..arg[1].__classname == "Entity" then
			self.__ptr = arg[1].__ptr
			return
		end
	end
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.SceneMesh(unpack(arg))
	end
end

function SceneMesh:SceneMeshFromMesh(mesh)
	local retVal = Polycode.SceneMesh_SceneMeshFromMesh(self.__ptr, mesh.__ptr)
	if retVal == nil then return nil end
	local __c = _G["SceneMesh"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:Render(buffer)
	local retVal = Polycode.SceneMesh_Render(self.__ptr, buffer.__ptr)
end

function SceneMesh:getShaderPass(index)
	local retVal = Polycode.SceneMesh_getShaderPass(self.__ptr, index)
	if retVal == nil then return nil end
	local __c = _G["ShaderPass"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:getNumShaderPasses()
	local retVal =  Polycode.SceneMesh_getNumShaderPasses(self.__ptr)
	return retVal
end

function SceneMesh:addShaderPass(pass)
	local retVal = Polycode.SceneMesh_addShaderPass(self.__ptr, pass.__ptr)
end

function SceneMesh:removeShaderPass(shaderIndex)
	local retVal = Polycode.SceneMesh_removeShaderPass(self.__ptr, shaderIndex)
end

function SceneMesh:getMesh()
	local retVal =  Polycode.SceneMesh_getMesh(self.__ptr)
	if retVal == nil then return nil end
	local __c = _G["shared_ptr<Mesh>"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:getMaterial()
	local retVal =  Polycode.SceneMesh_getMaterial(self.__ptr)
	if retVal == nil then return nil end
	local __c = _G["shared_ptr<Material>"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:loadSkeleton(fileName)
	local retVal = Polycode.SceneMesh_loadSkeleton(self.__ptr, fileName)
	if retVal == nil then return nil end
	local __c = _G["shared_ptr<Skeleton>"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:clearMaterial()
	local retVal =  Polycode.SceneMesh_clearMaterial(self.__ptr)
end

function SceneMesh:setMaterial(material)
	local retVal = Polycode.SceneMesh_setMaterial(self.__ptr, material.__ptr)
end

function SceneMesh:setMaterialByName(materialName, resourcePool)
	local retVal = Polycode.SceneMesh_setMaterialByName(self.__ptr, materialName, resourcePool.__ptr)
end

function SceneMesh:setMesh(mesh)
	local retVal = Polycode.SceneMesh_setMesh(self.__ptr, mesh.__ptr)
end

function SceneMesh:setSkeleton(skeleton)
	local retVal = Polycode.SceneMesh_setSkeleton(self.__ptr, skeleton.__ptr)
end

function SceneMesh:getSkeleton()
	local retVal =  Polycode.SceneMesh_getSkeleton(self.__ptr)
	if retVal == nil then return nil end
	local __c = _G["shared_ptr<Skeleton>"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:setLineWidth(newWidth)
	local retVal = Polycode.SceneMesh_setLineWidth(self.__ptr, newWidth)
end

function SceneMesh:getFilename()
	local retVal =  Polycode.SceneMesh_getFilename(self.__ptr)
	return retVal
end

function SceneMesh:setFilename(fileName)
	local retVal = Polycode.SceneMesh_setFilename(self.__ptr, fileName)
end

function SceneMesh:loadFromFile(fileName)
	local retVal = Polycode.SceneMesh_loadFromFile(self.__ptr, fileName)
end

function SceneMesh:customHitDetection(ray)
	local retVal = Polycode.SceneMesh_customHitDetection(self.__ptr, ray.__ptr)
	return retVal
end

function SceneMesh:setForceMaterial(forceMaterial)
	local retVal = Polycode.SceneMesh_setForceMaterial(self.__ptr, forceMaterial)
end

function SceneMesh:getForceMaterial()
	local retVal =  Polycode.SceneMesh_getForceMaterial(self.__ptr)
	return retVal
end

function SceneMesh:Clone(deepClone, ignoreEditorOnly)
	local retVal = Polycode.SceneMesh_Clone(self.__ptr, deepClone, ignoreEditorOnly)
	if retVal == nil then return nil end
	local __c = _G["Entity"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function SceneMesh:applyClone(clone, deepClone, ignoreEditorOnly)
	local retVal = Polycode.SceneMesh_applyClone(self.__ptr, clone.__ptr, deepClone, ignoreEditorOnly)
end

function SceneMesh:__delete()
	if self then Polycode.delete_SceneMesh(self.__ptr) end
end

File: bindings/javascript/Polycode/SceneMesh.js
function SceneMesh() {
	Object.defineProperties(this, {
		'lineWidth': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_lineWidth, set: SceneMesh.prototype.__set_lineWidth},
		'lineSmooth': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_lineSmooth, set: SceneMesh.prototype.__set_lineSmooth},
		'pointSmooth': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_pointSmooth, set: SceneMesh.prototype.__set_pointSmooth},
		'useGeometryHitDetection': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_useGeometryHitDetection, set: SceneMesh.prototype.__set_useGeometryHitDetection},
		'alphaTest': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_alphaTest, set: SceneMesh.prototype.__set_alphaTest},
		'backfaceCulled': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_backfaceCulled, set: SceneMesh.prototype.__set_backfaceCulled},
		'sendBoneMatricesToMaterial': { enumerable: true, configurable: true, get: SceneMesh.prototype.__get_sendBoneMatricesToMaterial, set: SceneMesh.prototype.__set_sendBoneMatricesToMaterial}
	})
}
SceneMesh.prototype.__get_lineWidth = function() {
	return Polycode.SceneMesh__get_lineWidth(this.__ptr)
}

SceneMesh.prototype.__set_lineWidth = function(val) {
	Polycode.SceneMesh__set_lineWidth(this.__ptr, val)
}

SceneMesh.prototype.__get_lineSmooth = function() {
	return Polycode.SceneMesh__get_lineSmooth(this.__ptr)
}

SceneMesh.prototype.__set_lineSmooth = function(val) {
	Polycode.SceneMesh__set_lineSmooth(this.__ptr, val)
}

SceneMesh.prototype.__get_pointSmooth = function() {
	return Polycode.SceneMesh__get_pointSmooth(this.__ptr)
}

SceneMesh.prototype.__set_pointSmooth = function(val) {
	Polycode.SceneMesh__set_pointSmooth(this.__ptr, val)
}

SceneMesh.prototype.__get_useGeometryHitDetection = function() {
	return Polycode.SceneMesh__get_useGeometryHitDetection(this.__ptr)
}

SceneMesh.prototype.__set_useGeometryHitDetection = function(val) {
	Polycode.SceneMesh__set_useGeometryHitDetection(this.__ptr, val)
}

SceneMesh.prototype.__get_alphaTest = function() {
	return Polycode.SceneMesh__get_alphaTest(this.__ptr)
}

SceneMesh.prototype.__set_alphaTest = function(val) {
	Polycode.SceneMesh__set_alphaTest(this.__ptr, val)
}

SceneMesh.prototype.__get_backfaceCulled = function() {
	return Polycode.SceneMesh__get_backfaceCulled(this.__ptr)
}

SceneMesh.prototype.__set_backfaceCulled = function(val) {
	Polycode.SceneMesh__set_backfaceCulled(this.__ptr, val)
}

SceneMesh.prototype.__get_sendBoneMatricesToMaterial = function() {
	return Polycode.SceneMesh__get_sendBoneMatricesToMaterial(this.__ptr)
}

SceneMesh.prototype.__set_sendBoneMatricesToMaterial = function(val) {
	Polycode.SceneMesh__set_sendBoneMatricesToMaterial(this.__ptr, val)
}


SceneMesh.prototype.Render = function(buffer) {
	Polycode.SceneMesh_Render(this.__ptr, buffer)
}

SceneMesh.prototype.getShaderPass = function(index) {
	var retVal = new ShaderPass()
	retVal.__ptr = Polycode.SceneMesh_getShaderPass(this.__ptr, index)
	return retVal
}

SceneMesh.prototype.getNumShaderPasses = function() {
	return Polycode.SceneMesh_getNumShaderPasses(this.__ptr)
}

SceneMesh.prototype.addShaderPass = function(pass) {
	Polycode.SceneMesh_addShaderPass(this.__ptr, pass)
}

SceneMesh.prototype.removeShaderPass = function(shaderIndex) {
	Polycode.SceneMesh_removeShaderPass(this.__ptr, shaderIndex)
}

SceneMesh.prototype.getMesh = function() {
	var retVal = new shared_ptr<Mesh>()
	retVal.__ptr = Polycode.SceneMesh_getMesh(this.__ptr)
	return retVal
}

SceneMesh.prototype.getMaterial = function() {
	var retVal = new shared_ptr<Material>()
	retVal.__ptr = Polycode.SceneMesh_getMaterial(this.__ptr)
	return retVal
}

SceneMesh.prototype.loadSkeleton = function(fileName) {
	var retVal = new shared_ptr<Skeleton>()
	retVal.__ptr = Polycode.SceneMesh_loadSkeleton(this.__ptr, fileName)
	return retVal
}

SceneMesh.prototype.clearMaterial = function() {
	Polycode.SceneMesh_clearMaterial(this.__ptr)
}

SceneMesh.prototype.setMaterial = function(material) {
	Polycode.SceneMesh_setMaterial(this.__ptr, material)
}

SceneMesh.prototype.setMaterialByName = function(materialName,resourcePool) {
	Polycode.SceneMesh_setMaterialByName(this.__ptr, materialName,resourcePool)
}

SceneMesh.prototype.setMesh = function(mesh) {
	Polycode.SceneMesh_setMesh(this.__ptr, mesh)
}

SceneMesh.prototype.setSkeleton = function(skeleton) {
	Polycode.SceneMesh_setSkeleton(this.__ptr, skeleton)
}

SceneMesh.prototype.getSkeleton = function() {
	var retVal = new shared_ptr<Skeleton>()
	retVal.__ptr = Polycode.SceneMesh_getSkeleton(this.__ptr)
	return retVal
}

SceneMesh.prototype.applySkeletonLocally = function() {
	Polycode.SceneMesh_applySkeletonLocally(this.__ptr)
}

SceneMesh.prototype.setLineWidth = function(newWidth) {
	Polycode.SceneMesh_setLineWidth(this.__ptr, newWidth)
}

SceneMesh.prototype.getFilename = function() {
	return Polycode.SceneMesh_getFilename(this.__ptr)
}

SceneMesh.prototype.setFilename = function(fileName) {
	Polycode.SceneMesh_setFilename(this.__ptr, fileName)
}

SceneMesh.prototype.loadFromFile = function(fileName) {
	Polycode.SceneMesh_loadFromFile(this.__ptr, fileName)
}

SceneMesh.prototype.customHitDetection = function(ray) {
	return Polycode.SceneMesh_customHitDetection(this.__ptr, ray)
}

SceneMesh.prototype.setForceMaterial = function(forceMaterial) {
	Polycode.SceneMesh_setForceMaterial(this.__ptr, forceMaterial)
}

SceneMesh.prototype.getForceMaterial = function() {
	return Polycode.SceneMesh_getForceMaterial(this.__ptr)
}

SceneMesh.prototype.Clone = function(deepClone,ignoreEditorOnly) {
	var retVal = new Entity()
	retVal.__ptr = Polycode.SceneMesh_Clone(this.__ptr, deepClone,ignoreEditorOnly)
	return retVal
}

SceneMesh.prototype.applyClone = function(clone,deepClone,ignoreEditorOnly) {
	Polycode.SceneMesh_applyClone(this.__ptr, clone,deepClone,ignoreEditorOnly)
}
